File: reference/original-java/src/hammerwatchgen/ScriptNode.java
package hammerwatchgen;

import java.util.ArrayList;

public class ScriptNode extends XMLObject {

    static ArrayList<ScriptNode> nodes = new ArrayList<>();
    
    public enum NodeType {
        
        ToggleElement( "ToggleElement" ),
        AreaTrigger( "AreaTrigger" ),
        RespawnPlayers( "RespawnPlayers" ),
        ShopArea( "ShopArea" ),
        LevelStart( "LevelStart" ),
        LevelExit( "LevelExitArea" ),
        AnnounceText( "AnnounceText" ),
        ObjectEventTrigger( "ObjectEventTrigger" ),
        RectangleShape( "RectangleShape" ),
        GameEnd( "GameEnd" );
        
        public XMLString type;
        NodeType( String name ) {
            
            this.type = new XMLString( "type", name );
        }
    }
    
    int id;
    boolean enabled;
    int triggerTimes;
    float x;
    float y;
    NodeType type;
    ArrayList<ScriptNode> connections;
    
    public static ScriptNode Create( float x, float y, NodeType type ) {
        
        ScriptNode n;
        switch( type ) {
            
            case LevelStart:
                n = new NodeLevelStart( x, y, type );
                break;
               
            case LevelExit:
                n = new NodeLevelExit( x, y, type );
                break;
                
            case RectangleShape:
                n = new NodeRectangleShape( x, y, type );
                break;
                
            case ShopArea:
                n = new NodeShopArea( x, y, type );
                break;

            case AnnounceText:
                n = new NodeAnnounceText( x, y, type );
                break;
   
            case ObjectEventTrigger:
                n = new NodeObjectEventTrigger( x, y, type );
                break;
          
            case ToggleElement:
                n = new NodeToggleElement( x, y, type );
                break;
                
            case AreaTrigger:
                n = new NodeAreaTrigger( x, y, type );
                break;
                
            case GameEnd:
            	n = new NodeGameEnd( x, y, type );
            	break;
                
            default:
                n = new ScriptNode( x, y, type );
                break;
        }
        nodes.add(n);
        return n;
    }
    
    public static void Delete( ScriptNode n ) {
        
        nodes.remove(n);
    }
    
    public static void Clear() {
        
        nodes.clear();
    }
    
    ScriptNode( float x, float y, NodeType type ) {
        
        connections = new ArrayList<>();
        this.x = x;
        this.y = y;
        this.type = type;
        this.triggerTimes = -1;
        this.enabled = true;
        this.id = Level.idCounter++;
    }
    
    public void connectTo( ScriptNode n ) {
        
        connections.add( n );
    }
    
    protected XMLDictionary getParametersDict() {
        
        return new XMLDictionary( "parameters" );
    }
    
    public String getXML() {
                
        // create XML structure
        XMLInt idInt = new XMLInt( "id", id );
        XMLString typeString = type.type;
        XMLBool enabledBool = new XMLBool( "enabled", enabled );
        XMLInt triggerInt = new XMLInt( "trigger-times", triggerTimes );
        XMLFloat xFloat = new XMLFloat( "x", x );
        XMLFloat yFloat = new XMLFloat( "y", y );
        XMLDictionary parameters = getParametersDict();
        
        XMLDictionary nodeDict = new XMLDictionary( "" );
        nodeDict.addData(idInt);
        nodeDict.addData(typeString);
        nodeDict.addData(enabledBool);
        nodeDict.addData(triggerInt);
        nodeDict.addData(xFloat);
        nodeDict.addData(yFloat);
        nodeDict.addData(parameters);

        if( !connections.isEmpty() ) {
            
            // create array
            int[] ids = new int[ connections.size() ];
            int[] delays = new int[ connections.size() ];
            for( int i = 0; i < ids.length; i++ ) {
                
                ids[i] = connections.get(i).id;
                delays[i] = 0;
            }
        XMLIntArray conArray = new XMLIntArray( "connections", ids );    
        XMLIntArray delayArray = new XMLIntArray( "delays", ids );
        nodeDict.addData( conArray );
        nodeDict.addData( delayArray );
        }
        
        return nodeDict.getXML();
    }
}

File: src/generator/objects/scriptNode.ts
import { XMLBool, XMLDictionary, XMLFloat, XMLInt, XMLIntArray, XMLObject, XMLRaw, XMLString } from '../xml'
import type { GenerationContext } from '../core/context'

export type NodeTypeName =
  | 'ToggleElement'
  | 'AreaTrigger'
  | 'RespawnPlayers'
  | 'ShopArea'
  | 'LevelStart'
  | 'LevelExitArea'
  | 'AnnounceText'
  | 'ObjectEventTrigger'
  | 'RectangleShape'
  | 'GameEnd'
  | 'SpawnObject'
  | 'GlobalEventTrigger'
  | 'TimerTrigger'
  | 'DestroyObject'
  | 'ToggleImmortality'
  | 'DangerArea'
  | 'PlaySound'
  | 'ChangeDoodadState'
  | 'ProjectileSpewer'

/**
 * Base scripting node (ported from ScriptNode.java). Most subclasses override
 * getParametersDict() to emit their type-specific `<dictionary name="parameters">`
 * block; a few (SpawnObject, GlobalEventTrigger, TimerTrigger) carry a bare
 * scalar instead and override getParametersXML() directly — see the seam below.
 */
export class ScriptNode extends XMLObject {
  id: number
  enabled = true
  triggerTimes = -1
  connections: ScriptNode[] = []

  /**
   * Per-connection delays in milliseconds, or `null` while every connection was
   * made without one — the overwhelmingly common case, and the one that has to
   * keep emitting the legacy `delays` line below byte-for-byte.
   */
  private delaysMs: number[] | null = null

  constructor(
    ctx: GenerationContext,
    public x: number,
    public y: number,
    public type: NodeTypeName
  ) {
    super()
    this.id = ctx.idCounter++
    ctx.scriptNodes.push(this)
  }

  /**
   * Connects to `n`, optionally after `delayMs` milliseconds.
   *
   * Passing a delay — at any point, even on the tenth connection — switches this
   * node into real-delay mode for good: connections already made are back-filled
   * with 0 and every later one records its own value (0 when omitted). A node
   * that is never given a delay stays in legacy mode and its XML is unchanged.
   */
  connectTo(n: ScriptNode, delayMs?: number): void {
    if (delayMs !== undefined && this.delaysMs === null) {
      this.delaysMs = this.connections.map(() => 0)
    }
    this.connections.push(n)
    if (this.delaysMs !== null) this.delaysMs.push(delayMs ?? 0)
  }

  protected getParametersDict(): XMLDictionary {
    return new XMLDictionary('parameters')
  }

  /**
   * The node's `parameters` element as raw XML.
   *
   * Most nodes emit a `<dictionary name="parameters">`, but SpawnObject,
   * GlobalEventTrigger and TimerTrigger carry a bare scalar instead
   * ([VERIFIED] 2026-08-10, see DISCOVERY-LOG.md), which a dictionary
   * cannot express — hence the seam.
   */
  protected getParametersXML(): string {
    return this.getParametersDict().getXML()
  }

  getXML(): string {
    const dict = new XMLDictionary('')
    dict.addData(new XMLInt('id', this.id))
    dict.addData(new XMLString('type', this.type))
    dict.addData(new XMLBool('enabled', this.enabled))
    dict.addData(new XMLInt('trigger-times', this.triggerTimes))
    dict.addData(new XMLFloat('x', this.x))
    dict.addData(new XMLFloat('y', this.y))
    dict.addData(new XMLRaw(this.getParametersXML()))

    if (this.connections.length > 0) {
      const ids = this.connections.map((c) => c.id)
      dict.addData(new XMLIntArray('connections', ids))
      if (this.delaysMs === null) {
        // the original passed the ids array for delays too — kept for output parity
        dict.addData(new XMLIntArray('delays', ids))
      } else {
        // Real delays go out under BOTH names. The generator has always written
        // `delays` (a verbatim id copy, which the engine evidently ignores or
        // rounds off to nothing visible), while the game's own editor writes
        // `connection-delays` with true milliseconds — see the DISCOVERY-LOG's
        // dialect note. Which key the engine actually honours in a *generated*
        // level is unverified, so a node that genuinely needs its timing to land
        // ships the same true values under each. [EMITTED] 2026-08-22
        dict.addData(new XMLIntArray('delays', this.delaysMs))
        dict.addData(new XMLIntArray('connection-delays', this.delaysMs))
      }
    }

    return dict.getXML()
  }
}
